### PR TITLE
A/V Sync using rtcp SR

### PR DESCRIPTION
**Pull request by kekkokk**
Made a better implementation of a/v sync for recordings based on sr rtcp packets.
tested a lot in these days and no problem yet.

It's my first ever code written in C and I really don't know where to start to implements test classes. If anyone could help me I'll be very grateful.

Works also when no audio is published and with OPUS/PCMU.

### CODE REVIEW

**Comment by jcague** [comment]
@kekkokk thanks a lot for the PR! there are some lint errors in this PR, can you please solve them first?

Travis shows them from this line on: https://travis-ci.org/ging/licode/builds/167089850#L7521

I'll check why it marks the build as green in the time being.

**Comment by kekkokk** [comment]
Done, I think

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 1]
```diff
@@ -327,18 +390,21 @@ int ExternalOutput::deliverAudioData_(char* buf, int len) {
 int ExternalOutput::deliverVideoData_(char* buf, int len) {
   if (videoSourceSsrc_ == 0) {
     RtpHeader* h = reinterpret_cast<RtpHeader*>(buf);
-    videoSourceSsrc_ = h->getSSRC();
+    if (h->getPayloadType() == RED_90000_PT || h->getPayloadType() == VP8_90000_PT) {
```
Comment (jnoring) [inline erizo/src/erizo/media/ExternalOutput.cpp]: It's probably worth discussing more with the project, but I think it makes sense to entirely disable FEC.  The interop issues between Firefox and Chrome are pretty significant.

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 2]
```diff
@@ -470,32 +548,39 @@ void ExternalOutput::queueData(char* buffer, int length, packetType type) {
...
     cond_.wait(lock);
-    while (audioQueue_.hasData()) {
-      boost::shared_ptr<dataPacket> audioP = audioQueue_.popPacket();
-      this->writeAudioData(audioP->data, audioP->length);
-    }
-    while (videoQueue_.hasData()) {
-      boost::shared_ptr<dataPacket> videoP = videoQueue_.popPacket();
-      this->writeVideoData(videoP->data, videoP->length);
+    if (hasAudio) {
+      while (audioQueue_.hasData() && audio_first_rtcp_time != 0 && video_first_rtcp_time != 0) {
```
Comment (jnoring) [inline erizo/src/erizo/media/ExternalOutput.cpp]: Can you abandon the outer `hasAudio` check and have the while loops simply look like `while (audioQueue_.hasData() && audio_first_rtcp_time != 0)` and `while (videoQueue.hasData() && video_first_rtcp_time != 0)`?

The one problem with this code is it only supports recording audio + video, and audio only--for our use case, we require external output to record video only (we have a phone bridge integration)

File: erizo/src/erizo/media/SyncUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    const double ntp_frac_ms = static_cast<double>(ntp_frac) / kNtpFracPerMs;
+    return ntp_secs * 1000 + ntp_frac_ms + 0.5;
+  }
+
+  // Utility function to avoid overflow (since rtpTimestamp is an uint32_t value)
+  // when getting the rtpTimestamp of an rtcp packet
+  long long SyncUtils::GetVectorPosition(int position, std::vector<RtcpHeader>& rtcp, bool isVideo) { //NOLINT
+    long long toReturn = rtcp[position].getRtpTimestamp();  //NOLINT
+    uint32_t firstRtpTimestamp = 0;
+    if (isVideo) {
```
Comment (jnoring) [inline erizo/src/erizo/media/SyncUtils.cpp]: Nit picker's corner; A little cleaner is:
```
uint32_t firstRtpTimestamp = isVideo ? firstVideoRtcpTimestamp : firstAudioRtcpTimestamp;
```

File: erizo/src/erizo/media/SyncUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      ELOG_DEBUG("Timestamp Wrapped - isVideo: %d, firstRTPTimestamp: %lld, wrapped: %lld",
+      isVideo, firstRtpTimestamp, toReturn);
+    }
+    return toReturn;
+  }
+
+  // Given a vector of rtcp packets and a timestamp, it finds the corresponding rtcp packet.
+  // It discard the old and unused rtcp from the vector and return true when finished or return false if the given rtp
+  // timestamp is < of the first rtcp timestamp
+  bool SyncUtils::Slide(long long rtp_timestamp, std::vector<RtcpHeader>& rtcp, bool isVideo) { //NOLINT
```
Comment (jnoring) [inline erizo/src/erizo/media/SyncUtils.cpp]: This seems like something that could use unit tests?

File: erizo/src/erizo/media/ExternalOutput.cpp (hunk 2 above)
Comment (kekkokk) [inline erizo/src/erizo/media/ExternalOutput.cpp]: It works video + audio or video only. I haven't tested audio only. Thanks for your precious suggestions, in the next days I'll try to adjust these things.

File: erizo/src/erizo/media/ExternalOutput.cpp (hunk 1 above)
Comment (jcague) [inline erizo/src/erizo/media/ExternalOutput.cpp]: @jnoring agree, we should make it configurable in anycase. But I saw at some place that those issues remain even with FEC disabled?

File: erizo/src/erizo/media/ExternalOutput.h [hunk 5]
```diff
@@ -94,7 +94,14 @@ class ExternalOutput : public MediaSink, public RawDataReceiver, public Feedback
   // composed of one or more partitions.  However, we don't seem to be sent anything but partition 0
   // so the second scheme seems not applicable.  Too bad.
   vp8SearchState vp8SearchState_;
-  bool needToSendFir_;
+  bool needToSendFir_, hasAudio, audioTimebaseSet;
```
Comment (jcague) [inline erizo/src/erizo/media/ExternalOutput.h]: please use `need_to_send_fir_`, `has_audio_`and, `audio_timebase_set` here. We're trying to follow Google's style guide from now on so double check new variables in this PR (we'll change the current code in future PRs)

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 6]
```diff
@@ -282,38 +306,77 @@ void ExternalOutput::writeVideoData(char* buf, int len) {
...
+      if (!SU.Slide(currentTimestamp, video_measurements.rtcp_Packets, true)) {
+        ELOG_WARN("Slide Video Failed, TS: %lld", currentTimestamp);
+        unpackagedSize_ = 0;
+        unpackagedBufferpart_ = unpackagedBuffer_;
+        return;
+      }
+      // Distance from rtp timestamp and the last rtcp timestamp
+      int delta_timestamp = currentTimestamp - video_measurements.rtcp_Packets[0].getRtpTimestamp();  // 1/90000
+      // Distance of the last rtcp packet and the first one
+      long long addend = av_rescale_q(SU.NtpToMs(video_measurements.rtcp_Packets[0].getNtpTimestampMSW(), //NOLINT
```
Comment (jcague) [inline erizo/src/erizo/media/ExternalOutput.cpp]: agree, it might help understanding this part

File: erizo/src/erizo/media/SyncUtils.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    // If the rtp timestamp is < of the first rtcp packet timestamp, return false.
+    // (discard this packet to avoid problems)
+    if (rtp_timestamp < GetVectorPosition(0, rtcp, isVideo)) {
+      ELOG_WARN("RtpToNtpMs: isVideo: %d, This packet is too early. Discard to avoid problems %lld / %lld, next: %lld",
+      isVideo, rtp_timestamp, GetVectorPosition(0, rtcp, isVideo), GetVectorPosition(1, rtcp, isVideo));
+      return false;
+    }
+    // Discard all the old rtcp packets.
+    while (rtcp.size() >1 && rtp_timestamp > GetVectorPosition(1, rtcp, isVideo)) {
```
Comment (jcague) [inline erizo/src/erizo/media/SyncUtils.cpp]: please use a helper function to discard old rtcp packets

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 8]
```diff
@@ -282,38 +306,77 @@ void ExternalOutput::writeVideoData(char* buf, int len) {
...
+                       video_measurements.rtcp_Packets[0].getNtpTimestampLSW()) -
+                       SU.NtpToMs(video_first_rtcp_ntp_timeMSW, video_first_rtcp_ntp_timeLSW), (AVRational){1, 1000},
```
Comment (jcague) [inline erizo/src/erizo/media/ExternalOutput.cpp]: consider using a variable to store `SU.NtpToMs(video_first_rtcp_ntp_timeMSW, video_first_rtcp_ntp_timeLSW)`, `SU.NtpToMs(video_measurements.rtcp_Packets[0].getNtpTimestampMSW(), video_measurements.rtcp_Packets[0].getNtpTimestampLSW())` and so on. 
it will help understanding the code (you won't need to comment it), and you won't need some NOLINT statements.

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 9]
```diff
@@ -282,38 +306,77 @@ void ExternalOutput::writeVideoData(char* buf, int len) {
...
+                       (AVRational){1, 90000});
+      long long offset = 0; //NOLINT
+      if (hasAudio) {
+        // Calculate the start offset of the A/V of the first rtcp packets using NTP timestamps.
+        offset = SU.NtpToMs(video_first_rtcp_ntp_timeMSW, video_first_rtcp_ntp_timeLSW) -
```
Comment (jcague) [inline erizo/src/erizo/media/ExternalOutput.cpp]: some of these NtpToMS values might be reused in this function

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 10]
```diff
@@ -154,26 +148,59 @@ void ExternalOutput::writeAudioData(char* buf, int len) {
...
+    ELOG_WARN("Slide Audio Failed, TS: %lld", currentTimestamp);
+    return;
+  }
 
-  // in practice, our timebase den is 1000, so this operation is a no-op.
-  timestampToWrite += audioOffsetMsec_ / (1000 / audio_stream_->time_base.den);
+  // Distance from rtp timestamp and the last rtcp timestamp
+  int delta_timestamp = currentTimestamp - audio_measurements.rtcp_Packets[0].getRtpTimestamp();
+  // Distance of the last rtcp packet and the first one
+  long long  addend = av_rescale_q(SU.NtpToMs(audio_measurements.rtcp_Packets[0].getNtpTimestampMSW(), //NOLINT
```
Comment (jcague) [inline erizo/src/erizo/media/ExternalOutput.cpp]: here you can also use variable to store temporary values. It's hard to follow the code with them and you have to use lots of NOLINT because of line lengths.

File: erizo/src/erizo/media/ExternalOutput.h [hunk 11]
```diff
@@ -94,7 +94,14 @@ class ExternalOutput : public MediaSink, public RawDataReceiver, public Feedback
...
+
+  Measurements video_measurements, audio_measurements;
```
Comment (jcague) [inline erizo/src/erizo/media/ExternalOutput.h]: please use a trailing underscores for private vars (`video_measurements_`)

File: erizo/src/erizo/media/ExternalOutput.h [hunk 12]
```diff
@@ -94,7 +94,14 @@ class ExternalOutput : public MediaSink, public RawDataReceiver, public Feedback
...
+  SyncUtils SU;
+  long long lastAudioTS, lastVideoTS; //NOLINT
```
Comment (jcague) [inline erizo/src/erizo/media/ExternalOutput.h]: consider `last_audio_ts_` and `last_video_ts_`

File: erizo/src/erizo/media/SyncUtils.cpp [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  // Converts an NTP timestamp to a millisecond timestamp.
+  int64_t SyncUtils::NtpToMs(uint32_t ntp_secs, uint32_t ntp_frac) {
+    const double ntp_frac_ms = static_cast<double>(ntp_frac) / kNtpFracPerMs;
+    return ntp_secs * 1000 + ntp_frac_ms + 0.5;
+  }
+
+  // Utility function to avoid overflow (since rtpTimestamp is an uint32_t value)
+  // when getting the rtpTimestamp of an rtcp packet
+  long long SyncUtils::GetVectorPosition(int position, std::vector<RtcpHeader>& rtcp, bool isVideo) { //NOLINT
+    long long toReturn = rtcp[position].getRtpTimestamp();  //NOLINT
```
Comment (jcague) [inline erizo/src/erizo/media/SyncUtils.cpp]: `to_return`, `is_video`, etc.

**Comment by kekkokk** [comment]
Awesome! I'll check as soon as possible

**Comment by zevarito** [comment]
@kekkokk will you re-open this one?

**Comment by kekkokk** [comment]
I don't know, actually I found a better way to implement av/sync with rtcp, and I also integrated the nack mechanism to prevent artifact caused by packet loss and also rtsp streaming.
The libary diverged a lot, so I need lot of time to divide my business logic from the licode vanilla.
Also is written soo bad (I admit, these are my first steps into c++ world) that i'm ashamed to pull request without refactoring.
As soon as I can I'll do a pr!
